feat(lsp): add textDocument/inlineCompletion support

devmanuelli opened the pull request:
Implements LSP 3.18 `textDocument/inlineCompletion` (ghost text).

## Demo

https://github.com/user-attachments/assets/8395cece-7992-4e1c-aa45-58c2893d6063

## Test

```bash
python3 -m venv .venv
source .venv/bin/activate
pip install pygls lsprotocol
```

**test_lsp.py:**
```python
#!/usr/bin/env python3
from lsprotocol import types as lsp
from pygls.lsp.server import LanguageServer

server = LanguageServer("test", "v0.1")

@server.feature(lsp.TEXT_DOCUMENT_INLINE_COMPLETION, lsp.InlineCompletionOptions())
def inline_completion(params: lsp.InlineCompletionParams):
    return lsp.InlineCompletionList(items=[
        lsp.InlineCompletionItem(insert_text="# Hello!")
    ])

if __name__ == "__main__":
    server.start_io()
```

**~/.config/helix/languages.toml:**
```toml
[language-server.test-lsp]
command = "/path/to/.venv/bin/python3"
args = ["/path/to/test_lsp.py"]

[[language]]
name = "python"
language-servers = ["test-lsp"]
```

**~/.config/helix/config.toml:**
```toml
[editor]
inline-completion-timeout = 150

[keys.insert]
tab = "inline_completion_accept"
C-e = "inline_completion_dismiss"
```

P.S.: *spread the word* on reddit!!

devmanuelli committed:
feat(lsp): add textDocument/inlineCompletion support

Implement LSP 3.18 inline completion (ghost text) feature:

- Add InlineCompletion to LanguageServerFeature enum
- Remove 'proposed' feature gate from lsp-types (official in LSP 3.18)
- Add inline_completion client capability and request method
- Store inline completion as InlineAnnotation on Document
- Render ghost text via text_annotations system
- Add inline_completion_accept and inline_completion_dismiss commands
- Auto-trigger on document change with configurable debounce (inline_completion_timeout)
- Clear completion on entering normal mode

devmanuelli commented:
I'm doing a refactoring simplifying the code

taylorplewe commented:
Nice!!! Can't wait to clone and try it out!

cunha commented:
Is this sufficient to get copilot-language-server working? If so, if someone configures it successfully, it would be useful to get a "bootstrap" config posted here.

(It seems copilot-language-server also requires other endpoints like `textDocument/didOpen`, `textDocument/didChange`, and  `textDocument/didClose`, but a quick search through the code indicates we may have these already.)

devmanuelli committed:
fix(lsp): handle inline completion range for correct ghost text display

- Add InlineCompletion struct to store display text, full insert text, and replace range
- Calculate offset from LSP range to show only new text as ghost text
- Use helix_core::Range directly instead of std::ops::Range
- Handle replace range on accept for proper text replacement
- Use safe slicing with get().unwrap_or_default()

synecdokey commented:
I would expect the cursor to stay where it is and the ghost text to appear after, no?

devmanuelli committed:
fix(lsp): render ghost text without shifting cursor position

Render inline completion ghost text directly at cursor position
instead of using InlineAnnotation, which was causing the cursor
to shift to the end of the ghost text. Also handles multiline
ghost text by rendering each line separately.

cunha commented:
Rocking this locally.  Works smoothly.  Also went through the code and don't have any suggestions.

Feature request: I would prefer to disable the automatic triggers for inline completions and instead have a keybind to trigger it on demand. 

I looked into this. It seems that the more involved part is passing the type of completion (autotrigger vs manual) to the event handler (the current code passes in `()`) to know when to ask for an inline completion (e.g., manual trigger with autotriggers disabled).  However, while working on a draft, I found that there already exists a `CompletionEvent` for basically the same purpose, and the `CompletionHandler::handle_event` function has logic that we might want when handling inline completions too (e.g., cancelling ongoing completion requests when the user deletes text).

Given I have very limited understanding of the overall architecture, I'm not sure what is the best approach.  I'm happy to work on it and submit a PR to https://github.com/devmanuelli/helix/tree/textDocument/inlineCompletion with some guidance.  I'm not even sure these are the right things to think about, but we would need to decide on whether to reuse the existing `CompletionEvent` structure (would require differentiating normal completions from inline completions, possibly touching the existing code for normal completions) or, in a more extreme case, to make inline completions a special case for the existing `CompletionHandler` to minimize code duplication.

devmanuelli committed:
feat(lsp): support multiple inline completions with cycling

- Query all LSP servers with inline completion capability, not just first
- Add InlineCompletions container with push/current/next/take_and_clear
- Add inline_completion_next command to cycle through completions
- Simplify accept logic using Transaction::change for both range cases
- Fix cursor position race by computing fresh cursor in response handler
- Add stale completion detection (discard if cursor moved past range)

devmanuelli commented:
> I would prefer to disable the automatic triggers for inline completions and instead have a keybind to trigger it on demand.

@cunha 
for each request or once for all the requests until u decide to disable it?

of course u can make PRs ;)

now I'm gonna push last commit where u can cycle among different completions from different LSPs

devmanuelli commented:
@ljahier I got a simpler way to solve it than yours. Probably InlineAnnotation field is not really needed

devmanuelli commented:
[](https://github.com/user-attachments/assets/18c4a14d-5783-404e-865e-31a57453fe84)

@ljahier

devmanuelli committed:
refactor(lsp): simplify InlineCompletion and fix non-ASCII handling

- Replace insert_text + offset with single ghost_text field (pre-computed)
- Make replace_range non-optional using Range::point for no-range case
- Use chars().skip() instead of byte slicing to handle non-ASCII correctly
- Use contains_range() for stale completion detection (cursor outside range)
- Simplify accept: insert ghost_text at cursor, delete up to range.to()

devmanuelli committed:
fix(lsp): add document version check for inline completion staleness

- Add doc_version check to detect content changes (e.g., select all + delete)
- Keep range check for cursor movement without content change (arrow keys)
- Simplify response handler using current!() instead of get_mut()
- Unify all staleness checks in one condition

devmanuelli committed:
feat(lsp): add manual trigger and auto-trigger config for inline completion

- Add `inline_completion_trigger` command for manual invocation
- Add `inline_completion_prev` command to cycle backwards
- Add `inline_completion_auto_trigger` config option (default: true)
- Extract trigger logic into public function for reuse
- Rename `next()` to `cycle(Direction)` for bidirectional support

devmanuelli commented on helix-term/src/handlers/inline_completion.rs:
```diff
@@ -0,0 +1,129 @@
+            };
+
+            let offset_encoding = ls.offset_encoding();
+            tokio::spawn(async move {
```
No it cannot happen that "stale requests might still update the UI" (look at lines 76-79 and 91-94)

devmanuelli committed:
feat(lsp): store all inline completion items from LSP response

Process all items returned by the LSP server instead of only the first one,
allowing users to cycle through multiple suggestions with next/prev commands.

devmanuelli commented:
@ljahier fix little bug (previously, we only took the first suggestion, even though the server could have sent us several ones)

cosmincartas commented:
I've tested it on my side and it looks good if the only LSP used is copilot. I'm getting some weird behavior when combining it with other LSPs such: "typescript-language-server" or "vue-language-server".

devmanuelli committed:
fix(ui): expand tabs in ghost text using document tab width

devmanuelli commented:
@the-mikedavis @pascalkuthe as u may already have vetted, my current implementation overwrites diagnostic text. Actually I would have found a solution w/ InlineAnnotation but cursor get pushed to the end. What should we prefer?

devmanuelli committed:
refactor(ui): integrate ghost text with annotation system

Switch inline completion ghost text from direct surface drawing to using
the text annotation and decoration systems. This allows proper coexistence
with diagnostics and other virtual text.

Changes:
- Mid-line: Use Overlay (first char) + InlineAnnotation (rest) to keep
  cursor in place while shifting diagnostics
- EOL: Use Decoration to render ghost text without shifting cursor,
  return column offset so diagnostics still shift
- Multi-line: Use LineAnnotation to reserve virtual lines + Decoration
  to render additional lines
- Add OnModeSwitch hook to clear completions when leaving insert mode
- Add documentation in docs/inline-completion-implementation.md

devmanuelli commented:
I'll keep this PR open but it will undergo major refactoring.
I don't give up I will bring Github copilot into Helix!!
The goal is to fully integrate it with other Helix's features like diagnostics (ATM we overwrite them ...)

[Screen Recording - Made with RecordCast (3).webm](https://github.com/user-attachments/assets/5065d4fa-65ef-4ea9-8efd-cc9cee2e5e48)

devmanuelli committed:
fix(ui): render EOL ghost text on cursor position with cursor style

- Fix ghost text rendering one column to the right of cursor at EOL
- Root cause: virt_off.col includes newline width (1), but cursor is ON
  the newline cell, not after it
- Solution: subtract 1 from virt_off.col for EOL ghost text positioning
- Also apply cursor style to first ghost character so it appears "on"
  the block cursor (matching mid-line behavior)
- Update documentation with fix details

devmanuelli committed:
fix(ui): use overlays for mid-line ghost text to prevent cursor shift

Replace InlineAnnotation with multiple Overlays for mid-line inline
completion display. This prevents the cursor from shifting when ghost
text is shown, providing a better editing experience.

- Use Overlay per character position (cursor to EOL) for mid-line case
- Add overflow text rendering via Decoration for chars beyond line end
- Apply suffix trimming only for display, keep ghost_text intact for accept
- Simplify Document cache to single overlay vector

devmanuelli committed:
fix(ui): dismiss inline completion on cursor movement

Clear ghost text when selection changes (e.g., arrow keys) to match
expected editor behavior where completions dismiss on cursor movement.

MohdShahulMalik commented:
Hey, do this with work lsp-ai language server as well? I tried to set up the inline completions with the lsp-ai language server but it doesn't seem to work. May someone help me with this please.

rohitbishnoi commented:
> Hey, do this with work lsp-ai language server as well? I tried to set up the inline completions with the lsp-ai language server but it doesn't seem to work. May someone help me with this please.

lsp-ai doesn't seems to be maintained anymore

devmanuelli committed:
fix(ui): improve multi-line ghost text rendering mid-line

For multi-line inline completions when cursor is mid-line:
- First line shown via overlays at cursor (no cursor shift)
- Additional lines shown as virtual lines below
- Original rest-of-line content pushed to bottom of ghost text

Single-line mid-line completions unchanged (overlay + overflow).

devmanuelli committed:
fix(ui): avoid duplicate rest-of-line in multi-line ghost text

Don't append rest_of_line to the last ghost text line if it already
appears anywhere in the ghost text (indicating it's part of the
replacement content rather than content to preserve).

archseer commented on helix-term/src/ui/editor.rs:
```diff
@@ -208,6 +208,23 @@ impl EditorView {
 
+        // Render ghost text at cursor position
+        if let Some(completion) = doc.inline_completions.current() {
+            if let Some(cursor_pos) = editor.cursor_cache.get(view, doc) {
```
Thanks for starting on this! To echo the comment on the linked issue, rather than modifying the rendering code we should be leveraging a mix of LineAnnotation (to modify the current line) and Decoration to display the virtual lines:

https://github.com/helix-editor/helix/issues/13039#issuecomment-2726028429

devmanuelli committed:
fix(ui): pad multi-line ghost text to cover entire rest-of-line

When displaying multi-line inline completions mid-line, the first line
preview was only creating overlays for as many characters as the ghost
text contained. If the ghost text first line was shorter than the
rest-of-line content, trailing characters would show through.

Fix by padding the preview with spaces to cover the full rest-of-line
length, ensuring all original content is blanked out (since it's being
pushed down to the last ghost text line).

Also removes the no-longer-needed implementation docs file.

leona commented:
Been waiting years for this. I hope it gets merged 🙏 Great work.

spion commented:
I've tested this locally and can confirm it works quite great with the real copilot language server. There may be some timing / location issues for suggested completions, and if you keep end-of-line diagnostics on as well it gets really croweded (I turned those off), but otherwise looking good! I have a forked branch with small tweaks to enable tab to accept completion (configurable). I think the fork will be my main driver from now on! :grinning: 

Awesome work. Thank you @devmanuelli

ewired commented:
I made another language server that is compatible with this PR: https://github.com/ewired/llama.lsp

It doesn't yet have all the context procurement features from https://github.com/ggml-org/llama.vim. Works fairly smooth with llama.cpp running GLM 4.7 Flash, which is normally a chat model, so infill is a bit lacking in intelligence.

babucarr32 commented:
It will be great to have this feature in helix.
